wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,185 @@ pub(crate) enum CommandEncoderStatus {
     ///
     /// [`Global::queue_submit`] drops command buffers unless they are
     /// in this state.
-    Finished,
+    Finished(CommandBufferMutable),
 
     /// An error occurred while recording a compute or render pass.
     ///
     /// When a `CommandEncoder` is left in this state, we have also
     /// returned an error result from the function that encountered
     /// the problem. Future attempts to use the encoder (for example,
-    /// calls to [`CommandBufferMutable::check_recording`]) will also return
-    /// errors.
-    ///
-    /// Calling [`Global::command_encoder_finish`] in this state
-    /// discards the command buffer under construction.
+    /// calls to [`Self::record`]) will also return errors.
     Error,
 }
 
+/// Changes the state of the given encoder.
+///
+/// Takes `&mut CommandEncoderStatus`, the previous state kind and the next state kind.
+///
+/// Returns `&mut CommandBufferMutable`.
+///
+/// Will panic if the previous state is not accurate.
+macro_rules! change_state_with_inner {
+    ($self:ident, $before:path, $after:path) => {{
+        let v = ::core::mem::replace($self, CommandEncoderStatus::Error);
+        let $before(inner) = v else {
+            unreachable!(
+                concat!(
+                    "internal error: command encoder was expected to be in ",
+                    stringify!($before),
+                    " state, but was {}"
+                ),
+                CommandEncoderStatus::display_state_discriminant($self)
+            )
+        };
+        let _ = ::core::mem::replace($self, $after(inner));
```

Comment:
What Erich said. If this can't be an assignment, it's not immediately clear to me why not. I'd like to see all the `let _ = replace` uses either commented, or replaced with a method call whose doc comment explains what's so delicate.